tests/deepspeed/test_deepspeed_multiple_model.py

```diff
@@ -0,0 +1,114 @@
+from accelerate.test_utils.training import RegressionDataset
+from accelerate.utils.deepspeed import DummyOptim, DummyScheduler, get_active_deepspeed_plugin
+
+
+GPT2_TINY = "sshleifer/tiny-gpt2"
+
+
+@require_deepspeed
+@require_non_cpu
+class DeepSpeedConfigIntegration(AccelerateTestCase):
+    def setUp(self):
+        super().setUp()
+
+        self.dist_env = dict(
+            ACCELERATE_USE_DEEPSPEED="true",
+            MASTER_ADDR="localhost",
+            MASTER_PORT="10999",
+            RANK="0",
+            LOCAL_RANK="0",
+            WORLD_SIZE="1",
+        )
+
+        self._test_file_path = inspect.getfile(self.__class__)
+        path = Path(self._test_file_path).resolve()
+        self.test_file_dir_str = str(path.parents[0])
+
+        self.ds_config_file = dict(
+            zero2=f"{self.test_file_dir_str}/ds_config_zero2.json",
+            zero3=f"{self.test_file_dir_str}/ds_config_zero3_model_only.json",
+        )
+
+        # use self.get_config_dict(stage) to use these to ensure the original is not modified
+        with open(self.ds_config_file["zero2"], encoding="utf-8") as f:
+            self.config_zero2 = json.load(f)
+        with open(self.ds_config_file["zero3"], encoding="utf-8") as f:
+            self.config_zero3 = json.load(f)
+
+        self.model_init = partial(AutoModelForCausalLM.from_pretrained, GPT2_TINY)
+
+    def get_ds_plugins(self):
+        return [
+            DeepSpeedPlugin(
+                hf_ds_config=self.config_zero2,
+            ),
+            DeepSpeedPlugin(
+                hf_ds_config=self.config_zero3,
+            ),
+        ]
+
+    def test_enable_disable(self):
+        ds_zero2, ds_zero3 = self.get_ds_plugins()
+        accelerator = Accelerator(
+            deepspeed_plugin=[ds_zero2, ds_zero3],
+        )
+        # Accelerator's constructor should automatically enable the first plugin
+        assert ds_zero2.enabled
+        assert not ds_zero3.enabled
+        assert get_active_deepspeed_plugin(accelerator.state) == ds_zero2
+        ds_zero3.enable()
+        assert not ds_zero2.enabled
+        assert ds_zero3.enabled
+        assert get_active_deepspeed_plugin(accelerator.state) == ds_zero3
+        ds_zero2.enable()
+        assert not ds_zero3.enabled
+        assert ds_zero2.enabled
+        assert get_active_deepspeed_plugin(accelerator.state) == ds_zero2
+
+    def test_prepare_multiple_models(self):
+        ds_zero2, ds_zero3 = self.get_ds_plugins()
+        accelerator = Accelerator(deepspeed_plugin=[ds_zero2, ds_zero3])
+        # Using Zero-2 first
+        model1 = self.model_init()
+        optimizer = DummyOptim(model1.parameters())
+        scheduler = DummyScheduler(optimizer)
+
+        dataset = RegressionDataset()
+        dataloader = torch.utils.data.DataLoader(dataset, batch_size=1)
+        model1, optimizer, scheduler, dataloader = accelerator.prepare(model1, optimizer, scheduler, dataloader)
+        ds_zero3.enable()
```

Comment:
do i understand correctly that calling this after `model1` is prepared will shard the model with z3? in other words, if i want to use z2 for model2 i would do something like:

```python
model1, optimizer, scheduler, dataloader = accelerator.prepare(model1, optimizer, scheduler, dataloader)
ds_zero3.enable()
model2 = self.model_init()
ds_zero2.enable()
```

Reply:
It will shard the following models with z2 if you call `.prepare`

For your case you'd do (realistically):

```python
# Zero2 requires real model and optimizer
ds_zero2.enable()
model1, optimizer, scheduler, dataloader = accelerator.prepare(model1, optimizer, scheduler, dataloader)
# Enable before model init so ZeroInit takes into effect
ds_zero3.enable()
model2 = self.model_init()
# Actually prepare with zero3
model2 = accelerator.prepare(model2)
```